File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.cs
public class Solution {
    private static string[] letterMap = new string[] {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    
    public IList<string> LetterCombinations(string digits) {
        var result = new List<string>();
        if (string.IsNullOrEmpty(digits)) return result;
        
        Backtrack(digits, 0, new char[digits.Length], result);
        return result;
    }
    
    private void Backtrack(string digits, int index, char[] current, List<string> result) {
        if (index == digits.Length) {
            result.Add(new string(current));
            return;
        }
        
        string letters = letterMap[digits[index] - '0'];
        for (int i = 0; i < letters.Length; i++) {
            current[index] = letters[i];
            Backtrack(digits, index + 1, current, result);
        }
    }
}

File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.dart
class Solution {
  static const List<String> letterMap = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
  
  List<String> letterCombinations(String digits) {
    List<String> result = [];
    if (digits.isEmpty) return result;
    
    void backtrack(int index, List<String> current) {
      if (index == digits.length) {
        result.add(current.join());
        return;
      }
      
      String letters = letterMap[int.parse(digits[index])];
      for (int i = 0; i < letters.length; i++) {
        current.add(letters[i]);
        backtrack(index + 1, current);
        current.removeLast();
      }
    }
    
    backtrack(0, []);
    return result;
  }
}